lib/persistence/mongo/Base.js

```diff
@@ -0,0 +1,115 @@
+var Q = require("q");
+var mongo_promise = require('./client.js');
+var crypto = require('crypto');
+var RecordNotFoundError = require('./RecordNotFound')
+
+module.exports = function(options){
+
+  var name = options['name'];
+  if ( !name ) { throw new Error("name is required"); }
+
+  var model = options['model'];
+  if ( !model ) { throw new Error("model is required"); }
+
+  var mongo_model = require('../backbone_sync.js')(model)
+
+  mongo_model = mongo_model.extend({
+
+    pk: function(){
+      return this.get('_id');
+    },
+
+    idAttribute: "_id",
+
+    create: function(attributes, options){
+
+      if(!this.collection_name){
+        throw new Error("need the collection_name attribute on the model!")
+      }
+
+      // 24 character id string, as opposed to using mongo's builtin ObjectID
+      this.set('_id', crypto.randomBytes(12).toString('hex'))
```

Comment:
I'm not really familiar with mongo, but it seems like we're breaking a
convention here, and I'm just trying to decide if it's the kind of
convention that we should be sticking to.

On Wed, Feb 4, 2015 at 10:13 AM, Sam Vevang <notifications@github.com>
wrote:

> In lib/persistence/mongo/Base.js
> <https://github.com/michaelkirk/weiqi/pull/28#discussion_r24104992>:
>
> > +  mongo_model = mongo_model.extend({
> > +
> > +    pk: function(){
> > +      return this.get('_id');
> > +    },
> > +
> > +    idAttribute: "_id",
> > +
> > +    create: function(attributes, options){
> > +
> > +      if(!this.collection_name){
> > +        throw new Error("need the collection_name attribute on the model!")
> > +      }
> > +
> > +      // 24 character id string, as opposed to using mongo's builtin ObjectID
> > +      this.set('_id', crypto.randomBytes(12).toString('hex'))
>
> ahh, and also you could easily predict the next id in the sequence. The
> boards were being created so quickly that the timestamps overlapped:
>
>   weiqi> console.log(ObjectID(), ObjectID())
> 54d26136aaebe2dc0b0420da 54d26136aaebe2dc0b0420db
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/michaelkirk/weiqi/pull/28/files#r24104992>.
>